types/index.d.ts

```diff
@@ -7,14 +7,17 @@ import {Node} from 'unist'
 declare namespace rehypeReact {
   type FragmentLike<T> = (props: any) => T | null
 
-  interface ComponentProps {
-    [prop: string]: unknown
-    node?: Node
+  type ComponentPropsWithoutNode = Record<string, unknown>
+
+  interface ComponentPropsWithNode extends ComponentPropsWithoutNode {
+    node: Node
   }
 
-  type ComponentLike<T> = (props: ComponentProps) => T | null
+  type ComponentProps = ComponentPropsWithoutNode | ComponentPropsWithNode
+
+  type ComponentLike<T, P = ComponentPropsWithNode> = (props: P) => T | null
```

Comment:
an unconstrained generic is roughly equivalent to `any`, it doesn't check anything

For example

```typescript
const a: rehypeToReact.ComponentLike<
  ReturnType<typeof React.createElement>,
  // this should require runtime narrowing, but does not
  { notARealProp: Symbol }
> = ({ notARealProp }: { notARealProp: Symbol }) => <>example</>;
```

is valid

this should probably be more along the lines of

```suggestion
type ComponentLike<T, P extends ComponentPropsWithNode = ComponentPropsWithNode> = (props: P) => T | null
```